src/connections.tsx

```diff
@@ -1,45 +1,177 @@
+  WebUsbChooseMicrobit,
+  ConnectingBluetooth,
+  ConnectingMicrobits,
+  FlashingInProgress,
+
+  // Failure stages
+  TryAgainReplugMicrobit,
+  TryAgainCloseTabs,
+  TryAgainSelectMicrobit,
+  TryAgainBluetoothConnect,
+  BadFirmware,
+  MicrobitUnsupported,
+  WebUsbBluetoothUnsupported,
+}
+
+export enum ConnType {
+  Bluetooth,
+  RadioBridge,
+  RadioRemote,
+}
+
+// For now, we only have one program type, but here we can expand to other
+// programs e.g. CollectDataInField
+export enum ProgramType {
+  CollectData,
+}
+
+export enum ConnStatus {
+  Disconnected,
+  Connected,
+}
+
+export interface BluetoothConn {
+  status: ConnStatus;
+  program: ProgramType;
+  type: ConnType.Bluetooth;
+  bluetoothPattern: boolean[];
+}
+
+interface RadioConn {
+  status: ConnStatus;
+  program: ProgramType;
+  type: Exclude<ConnType, ConnType.Bluetooth>;
+}
+
+export type Conn = BluetoothConn | RadioConn;
+
+export type ConnState = {
+  stage: ConnStage;
+  type: ConnType;
+  program: ProgramType;
+  isWebUsbSupported: boolean;
+  isWebBluetoothSupported: boolean;
+  connections: Conn[];
+};
+
+type ConnContextValue = {
   state: ConnState;
-  dispatch: React.Dispatch<ConnEvent>;
+  setState: (state: ConnState) => void;
 };
 
-export const ConnectionFlowContext =
-  createContext<ConnectionFlowContextValue | null>(null);
+export const ConnectionContext = createContext<ConnContextValue | null>(null);
 
-interface ConnectionFlowProviderProps {
+interface ConnectionProviderProps {
   children: ReactNode;
 }
 
-export const ConnectionFlowProvider = ({
-  children,
-}: ConnectionFlowProviderProps) => {
-  const isBluetoothSupported = true;
-  const [state, dispatch] = useReducer(connectionDialogReducer, {
-    // TODO: Check bt and usb compatibility
-    type: isBluetoothSupported ? ConnType.Bluetooth : ConnType.RadioRemote,
+export const isValidStoredConnections = (
+  v: unknown
+): v is StoredConnections => {
+  if (typeof v !== "object") {
+    return false;
+  }
+  const valueObject = v as object;
+  if (!("connections" in valueObject)) {
+    return false;
+  }
+  const conns = valueObject.connections;
+  if (!isArray(conns)) {
+    return false;
+  }
+  const array = conns as unknown[];
+  for (const item of array) {
+    if (typeof item !== "object" || item === null) {
+      return false;
+    }
+    if (!("program" in item) || !("type" in item) || !("status" in item)) {
+      return false;
+    }
+  }
+  return true;
+};
+
+interface StoredConnections {
+  connections: Conn[];
+}
+
+export const ConnectionProvider = ({ children }: ConnectionProviderProps) => {
+  // TODO: Check bt and usb compatibility
+  const isWebBluetoothSupported = true;
+  const isWebUsbSupported = true;
+
+  // Connections are stored in local storage to save bluetooth pattern
+  const [storedConnections, setStoredConnections] =
+    useStorage<StoredConnections>(
+      "local",
+      "connections",
+      { connections: [] },
+      isValidStoredConnections
+    );
```

Comment:
Not sure if I am storing too much in local storage just for the bluetooth pattern